src/v/coproc/router_source_manager.cc

```diff
@@ -0,0 +1,122 @@
+    // For now only support the 'latest' policy
+    if (!is_valid_ingestion_policy(src.tip)) {
+        src.promise.set_value(errc::invalid_ingestion_policy);
+        return;
+    }
+    absl::flat_hash_map<model::ntp, storage::log> logs
+      = _api.local().log_mgr().get(src.tn);
+    if (logs.empty()) {
+        src.promise.set_value(errc::topic_does_not_exist);
+        return;
+    }
+
+    for (const auto& [ntp, log] : logs) {
+        auto found = _sources.find(ntp);
+        if (found == _sources.end()) {
+            auto ts = ss::make_lw_shared<topic_state>(
+              topic_state{.log = log, .scripts = {id}});
+            _sources.emplace(ntp, std::move(ts));
+        } else {
+            found->second->scripts.emplace(id);
+        }
+        vlog(coproclog.info, "Inserted ntp {} id {}", ntp, id);
+    }
+    src.promise.set_value(errc::success);
+}
+
+ss::future<> router_source_manager::process_removals() {
+    return ss::do_with(std::move(_removals), [this](deferred_removals& drm) {
+        return ss::do_for_each(
+          drm.begin(), drm.end(), [this](deferred_removals::value_type& e) {
+              do_process_removals(e.first, e.second);
+          });
+    });
+}
+
+void router_source_manager::do_process_removals(
+  script_id id, ss::promise<bool>& promise) {
+    absl::flat_hash_set<model::ntp> deleted;
+    std::for_each(
+      _sources.begin(),
+      _sources.end(),
+      [&deleted, id = id](consumers_state::value_type& p) {
+          auto& scripts = p.second->scripts;
+          scripts.erase(id);
+          vlog(coproclog.info, "Deleted script id: {}", id);
+          if (scripts.empty()) {
+              deleted.emplace(p.first);
+          }
+      });
+    // If no more scripts are tracking an ntp, remove the ntp
+    absl::erase_if(_sources, [&deleted](const consumers_state::value_type& p) {
+        return deleted.contains(p.first);
+    });
+
+    promise.set_value(!deleted.empty());
+}
+
+ss::future<> router_source_manager::cancel_pending_updates() {
+    auto cda = ss::do_with(std::move(_additions), [](deferred_additions& dfa) {
+        return ss::parallel_for_each(
+          dfa.begin(), dfa.end(), [](deferred_additions::value_type& src) {
+              src.second.promise.set_exception(
+                std::make_exception_ptr(std::runtime_error(fmt_with_ctx(
+                  fmt::format, "Pending addition cancelled: {}", src.first))));
+          });
+    });
+    auto cdc = ss::do_with(std::move(_removals), [](deferred_removals& drm) {
+        return ss::parallel_for_each(
+          drm.begin(), drm.end(), [](deferred_removals::value_type& r) {
+              r.second.set_exception(
+                std::make_exception_ptr(std::runtime_error(fmt_with_ctx(
+                  fmt::format, "Pending removal cancelled: {}", r.first))));
+          });
+    });
+    return ss::when_all(std::move(cda), std::move(cdc))
```

Comment:
@senior7515 is this what you were thinking of when you mentioned that we should add exception handling here?

Reply:
question, why do any of these have to be a future in the first place ? 

`.set_exception` is not a future ? why not just for loop ?

Reply:
setting the value does not yield. just sets the value. what's the point in making this `future`. 

furthermore, what's the point in parallel. you can probably iterate faster through the loop than creating a bunch of heap structures

Reply:
good point, no need for futures here